File: 0525-contiguous-array/0525-contiguous-array.java
class Solution {
    public int findMaxLength(int[] nums) {
        HashMap<Integer, Integer>mp = new HashMap<>();
        int ans = 0, sum = 0;
        mp.put(0, -1);
        for(int i=0; i<nums.length; i++) {
            sum += nums[i] == 0 ? -1 : 1;
            if(mp.containsKey(sum)) {
                ans = Math.max(ans, i - mp.get(sum));
            }
            else {
                mp.put(sum, i);
            }
        }
        return ans;
    }
}

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int>mp;
        int n = nums.size();
        int sum = 0, ans = 0;
        mp[sum] = -1;
        for(int i=0; i<n; i++) {
            int x = nums[i] == 0 ? -1 : 1;
            sum += x;
            if(mp.find(sum) != mp.end()) {
                ans = max(ans, i - mp[sum]);
            }
            else {
                mp[sum] = i;
            }
        }
        return ans;
    }
};